sway/ipc-json.c

```diff
@@ -0,0 +1,326 @@
+#include "container.h"
+#include "util.h"
+#include "ipc-json.h"
+
+static json_object *ipc_json_create_rect(swayc_t *c) {
+	json_object *rect = json_object_new_object();
+
+	json_object_object_add(rect, "x", json_object_new_int((int32_t)c->x));
+	json_object_object_add(rect, "y", json_object_new_int((int32_t)c->y));
+	json_object_object_add(rect, "width", json_object_new_int((int32_t)c->width));
+	json_object_object_add(rect, "height", json_object_new_int((int32_t)c->height));
+
+	return rect;
+}
+
+static const char *ipc_json_border_description(swayc_t *c) {
+	const char *border;
+
+	switch (c->border_type) {
+	case B_PIXEL:
+		border = "1pixel";
+		break;
+
+	case B_NORMAL:
+		border = "normal";
+		break;
+
+	case B_NONE: // fallthrough
+	default:
+		border = "none";
+		break;
+	}
+
+	return border;
+}
+
+static const char *ipc_json_layout_description(swayc_t *c) {
+	const char *layout;
+
+	switch (c->layout) {
+	case L_VERT:
+		layout = "splitv";
+		break;
+
+	case L_HORIZ:
+		layout = "splith";
+		break;
+
+	case L_TABBED:
+		layout = "tabbed";
+		break;
+
+	case L_STACKED:
+		layout = "stacked";
+		break;
+
+	case L_FLOATING:
+		layout = "floating";
+		break;
+
+	case L_NONE: // fallthrough
+	case L_LAYOUTS: // fallthrough; this should never happen, I'm just trying to silence compiler warnings
+	default:
+		layout = "null";
+		break;
+	}
+
+	return layout;
+}
+
+static float ipc_json_child_percentage(swayc_t *c) {
+	float percent = 0;
+	swayc_t *parent = c->parent;
+
+	if (parent) {
+		switch (parent->layout) {
+		case L_VERT:
+			percent = c->height / parent->height;
+			break;
+
+		case L_HORIZ:
+			percent = c->width / parent->width;
+			break;
+
+		case L_STACKED: // fallthrough
+		case L_TABBED: // fallthrough
+			percent = 1.0;
+			break;
+
+		default:
+			break;
+		}
+	}
+
+	return percent;
+}
+
+static void ipc_json_describe_output(swayc_t *output, json_object *object) {
+	json_object_object_add(object, "active", json_object_new_boolean(true));
+	json_object_object_add(object, "primary", json_object_new_boolean(false));
+	json_object_object_add(object, "layout", json_object_new_string("output"));
+	json_object_object_add(object, "type", json_object_new_string("output"));
+	json_object_object_add(object, "current_workspace",
+		(output->focused) ? json_object_new_string(output->focused->name) : NULL);
+}
+
+static void ipc_json_describe_workspace(swayc_t *workspace, json_object *object) {
+	int num = (isdigit(workspace->name[0])) ? atoi(workspace->name) : -1;
+	bool focused = root_container.focused == workspace->parent && workspace->parent->focused == workspace;
+
+	json_object_object_add(object, "num", json_object_new_int(num));
+	json_object_object_add(object, "visible", json_object_new_boolean(workspace->visible));
+	json_object_object_add(object, "focused", json_object_new_boolean(focused));
+	json_object_object_add(object, "output", json_object_new_string((workspace->parent) ? workspace->parent->name : "null"));
```

Comment:
That was in the `get_workspace()` code originally and I followed that in a couple other places too, but I don't really know what would be most appropriate. The [JSON spec](http://www.json.org/) indicates `null` is a valid option so I could change it to that for all cases if preferable.